Simplify and extend broadcast eltype promotion mechanism

Sacha0 opened the pull request:
(Requires #19667.) This pull request restores the simplification of `broadcast`'s eltype promotion mechanism in #19421. With benefit of #19667 (thanks `@yuyichao`!), that simplification handles more cases (e.g. closures accepting more than two arguments) than the present mechanism (`_default_eltype`-`first`-`Generator`-`zip`). This pull request also gives `broadcast`'s eltype promotion mechanism a more precise name (`_broadcast_type` -> `_broadcast_eltype`). Best!

pabloferz commented on base/broadcast.jl:
```diff
@@ -269,22 +269,14 @@ end
-
-_broadcast_type(f, T::Type, As...) = Base._return_type(f, typestuple(T, As...))
-_broadcast_type(f, A, Bs...) = Base._default_eltype(Base.Generator{ziptype(A, Bs...), ftype(f, A, Bs...)})
+eltypestuple(a) = (Base.@_pure_meta; Tuple{eltype(a)})
```
What do you think of moving these either to `abstractarray.jl` or `promotion.jl`?

Sacha0 replied:
Moving `eltypestuple` to `abstractarray.jl` or `promotion.jl`? `_broadcast_eltype` being the only consumer of `eltypestuple`, wherever `_broadcast_eltype` lives seems like the appropriate location? Or do you have other (broader) uses in mind for `eltypesuple`? Or did you mean something else altogether? Thanks for reviewing! :)

pabloferz replied:
Sorry for commenting above without providing a context. I was thinking that we could change the `promote_op` methods to take a function and the tuple of the `eltype`s and remove the `promote_op(Any...) = Any`, so it would make sense having `eltypestuple` somewhere in `Base`

Sacha0 replied:
Sounds like a good plan! Agreed another PR might be better. Perhaps we can chat about cleaning up these mechanisms after feature freeze? Thanks again!

Sacha0 commented:
Rewritten with the changes to `_broadcast_(el)type` in #16961 factored in. Unfortunately the additional parameter to `_broadcast_(el)type`/`eltypestuple` introduced by #16961 seems to kill this simplified mechanism's ability to handle the test case from #19421 without special casing (and generalizations of that case probably fail now as well). (Foisting that additional complication on the general `_broadcast_(el)type` mechanism to handle the idiosyncrasies of a particular type seems dubious.) Best!

Sacha0 commented:
(Pinging @pabloferz for review. Thanks in advance!)

tkelman commented on base/broadcast.jl:
```diff
@@ -282,32 +282,27 @@ end
-# other times, so there are two variants of typestuple.
+# _broadcast_eltype is broadcast's primary result-eltype promotion mechanism.
+# _broadcast_eltype uses eltypestuple to construct a tuple type of the eltypes
+# of the input-array arguments passed to _broadcast_eltype (from an upstream broacast).
```
broacast typo

Sacha0 replied:
That word. Fixed. Thanks!

Sacha0 commented:
Rebased and fixed. Collides with #19787, not certain which should take precedence. Best!

pabloferz commented:
If no one objects I'd prefer to merge that one first.

Sacha0 commented:
> If no one objects I'd prefer to merge that one first.

If we can merge #19787 within the next few hours, then #19787 should take precedence. Otherwise I advocate for merging this pull request first, as #19724 depends on this pull request. Best!

pabloferz commented:
I had forgotten about #19724. I agree with @Sacha0's proposal above.

Sacha0 committed:
Simplify broadcast's eltype promotion mechanism and make it handle more cases.

Re-simplify broadcast's eltype promotion mechanism as in #19421. With benefit of #19667, this simplified mechanism should handle additional cases (e.g. closures accepting more than two arguments). Also rename the mechanism more precisely (_broadcast_type -> _broadcast_eltype, typestuple -> eltypestuple).

Sacha0 commented:
Revised with #19787 in. Handles the test case from #19421 again without special casing. Best!

tkelman commented:
i686 travis failure looks related to https://github.com/JuliaLang/julia/issues/19792, most likely pre existing.
